File: web/src/IgnoreParser.ts
import { readFile } from "node:fs/promises";
import { join, relative } from "node:path";

interface IgnorePattern {
	pattern: string;
	isNegated: boolean;
	isDirectoryOnly: boolean;
	pathPrefix: string;
}

export class IgnoreParser {
	private patterns: Map<string, IgnorePattern[]> = new Map();

	async loadGitignore(repoPath: string): Promise<void> {
		this.patterns.clear();
		await this.loadGitignoreRecursive(repoPath, repoPath);
	}

	private async loadGitignoreRecursive(repoPath: string, currentDir: string): Promise<void> {
		const gitignorePath = join(currentDir, ".gitignore");
		const relativeDir = relative(repoPath, currentDir) || ".";

		try {
			const content = await readFile(gitignorePath, "utf-8");
			const patterns = this.parseGitignore(content, relativeDir);
			if (patterns.length > 0) {
				this.patterns.set(relativeDir, patterns);
			}
		} catch {
			// No .gitignore file in this directory
		}

		// Recursively check subdirectories (but skip .git)
		const { readdirSync, statSync } = await import("node:fs");
		try {
			const entries = readdirSync(currentDir);
			for (const entry of entries) {
				if (entry === ".git") continue;
				const fullPath = join(currentDir, entry);
				try {
					const stat = statSync(fullPath);
					if (stat.isDirectory()) {
						await this.loadGitignoreRecursive(repoPath, fullPath);
					}
				} catch {
					// Skip if can't stat
				}
			}
		} catch {
			// Skip if can't read directory
		}
	}

	private parseGitignore(content: string, pathPrefix: string): IgnorePattern[] {
		const patterns: IgnorePattern[] = [];
		const lines = content.split("\n");

		for (let line of lines) {
			line = line.trim();

			// Skip empty lines and comments
			if (!line || line.startsWith("#")) continue;

			// Handle negation (!)
			const isNegated = line.startsWith("!");
			if (isNegated) {
				line = line.slice(1);
			}

			// Handle directory-only patterns (trailing /)
			const isDirectoryOnly = line.endsWith("/");
			if (isDirectoryOnly) {
				line = line.slice(0, -1);
			}

			// Skip empty pattern after processing
			if (!line) continue;

			patterns.push({
				pattern: line,
				isNegated,
				isDirectoryOnly,
				pathPrefix,
			});
		}

		return patterns;
	}

	isIgnored(filePath: string, isDirectory = false): boolean {
		// Normalize path
		const normalizedPath = filePath.replace(/\\/g, "/");
		const pathParts = normalizedPath.split("/");

		// Check all applicable .gitignore files
		let ignored = false;

		for (const [, patterns] of this.patterns) {
			for (const pattern of patterns) {
				if (this.matchesPattern(normalizedPath, pathParts, pattern, isDirectory)) {
					ignored = !pattern.isNegated;
				}
			}
		}

		return ignored;
	}

	private matchesPattern(
		filePath: string,
		pathParts: string[],
		pattern: IgnorePattern,
		isDirectory: boolean,
	): boolean {
		// Check if pattern applies to this file based on path prefix
		if (pattern.pathPrefix !== ".") {
			const prefixParts = pattern.pathPrefix.split("/");
			if (
				pathParts.length < prefixParts.length ||
				!prefixParts.every((part, i) => part === pathParts[i])
			) {
				return false;
			}
		}

		// Get the relative path from the .gitignore location
		const relativePath =
			pattern.pathPrefix === "." ? filePath : filePath.slice(pattern.pathPrefix.length + 1);

		// If directory-only pattern, only match directories
		if (pattern.isDirectoryOnly && !isDirectory) {
			return false;
		}

		return this.matchPatternString(relativePath, pathParts, pattern.pattern);
	}

	private matchPatternString(filePath: string, pathParts: string[], pattern: string): boolean {
		// Convert gitignore pattern to regex
		let regexPattern = pattern;

		// Handle patterns with /
		if (pattern.includes("/")) {
			// Pattern with / is anchored
			if (pattern.startsWith("/")) {
				regexPattern = pattern.slice(1);
			}
		} else {
			// Pattern without / matches at any depth
			// Match against the last part or the whole path with wildcards
			const escapedPattern = this.escapeRegex(pattern).replace(/\*\*/g, ".*");
			const regex = new RegExp(
				`(^|/)${escapedPattern.replace(/\\\*/g, "[^/]*").replace(/\\\?/g, ".")}$`,
			);
			return regex.test(filePath);
		}

		// Handle ** (matches zero or more directories)
		regexPattern = regexPattern.replace(/\*\*/g, "<<<DOUBLESTAR>>>");

		// Handle * (matches anything except /)
		regexPattern = regexPattern.replace(/\*/g, "[^/]*");

		// Handle ? (matches single character except /)
		regexPattern = regexPattern.replace(/\?/g, "[^/]");

		// Restore ** as .*
		regexPattern = regexPattern.replace(/<<<DOUBLESTAR>>>/g, ".*");

		// Handle character classes [abc]
		regexPattern = regexPattern.replace(/\[([^\]]+)\]/g, "[$1]");

		// Escape other regex special characters
		regexPattern = this.escapeRegexExcept(regexPattern, "[^/].*$+()|");

		// Match pattern
		const regex = new RegExp(`^${regexPattern}(/.*)?$`);
		return regex.test(filePath);
	}

	private escapeRegex(str: string): string {
		return str.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
	}

	private escapeRegexExcept(str: string, except: string): string {
		const exceptSet = new Set(except);
		let result = "";
		for (const char of str) {
			if (".*+?^${}()|[]\\".includes(char) && !exceptSet.has(char)) {
				result += "\\" + char;
			} else {
				result += char;
			}
		}
		return result;
	}
}

File: dotnet/Gitologist/IgnoreParser.cs
namespace Gitologist;

public class IgnorePattern
{
    public string Pattern { get; set; } = string.Empty;
    public bool IsNegated { get; set; }
    public bool IsDirectoryOnly { get; set; }
    public string PathPrefix { get; set; } = string.Empty;
}

public class IgnoreParser
{
    private Dictionary<string, List<IgnorePattern>> _patterns = new();

    public async Task LoadGitignore(string repoPath)
    {
        _patterns.Clear();
        await LoadGitignoreRecursive(repoPath, repoPath);
    }

    private async Task LoadGitignoreRecursive(string repoPath, string currentDir)
    {
        var gitignorePath = Path.Combine(currentDir, ".gitignore");
        var relativeDir = GetRelativePath(repoPath, currentDir) ?? ".";

        try
        {
            var content = await File.ReadAllTextAsync(gitignorePath);
            var patterns = ParseGitignore(content, relativeDir);
            if (patterns.Count > 0)
            {
                _patterns[relativeDir] = patterns;
            }
        }
        catch
        {
            // No .gitignore file in this directory
        }

        // Recursively check subdirectories (but skip .git)
        try
        {
            var entries = Directory.GetFileSystemEntries(currentDir);
            foreach (var entry in entries)
            {
                var entryName = Path.GetFileName(entry);
                if (entryName == ".git") continue;

                if (Directory.Exists(entry))
                {
                    await LoadGitignoreRecursive(repoPath, entry);
                }
            }
        }
        catch
        {
            // Skip if can't read directory
        }
    }

    private List<IgnorePattern> ParseGitignore(string content, string pathPrefix)
    {
        var patterns = new List<IgnorePattern>();
        var lines = content.Split('\n');

        foreach (var line in lines)
        {
            var trimmedLine = line.Trim();

            // Skip empty lines and comments
            if (string.IsNullOrEmpty(trimmedLine) || trimmedLine.StartsWith("#"))
                continue;

            // Handle negation (!)
            var isNegated = trimmedLine.StartsWith("!");
            if (isNegated)
            {
                trimmedLine = trimmedLine.Substring(1);
            }

            // Handle directory-only patterns (trailing /)
            var isDirectoryOnly = trimmedLine.EndsWith("/");
            if (isDirectoryOnly)
            {
                trimmedLine = trimmedLine.Substring(0, trimmedLine.Length - 1);
            }

            // Skip empty pattern after processing
            if (string.IsNullOrEmpty(trimmedLine))
                continue;

            patterns.Add(new IgnorePattern
            {
                Pattern = trimmedLine,
                IsNegated = isNegated,
                IsDirectoryOnly = isDirectoryOnly,
                PathPrefix = pathPrefix
            });
        }

        return patterns;
    }

    public bool IsIgnored(string filePath, bool isDirectory = false)
    {
        var normalizedPath = filePath.Replace("\\", "/");
        var pathParts = normalizedPath.Split('/');

        var ignored = false;

        foreach (var patternList in _patterns.Values)
        {
            foreach (var pattern in patternList)
            {
                if (MatchesPattern(normalizedPath, pathParts, pattern, isDirectory))
                {
                    ignored = !pattern.IsNegated;
                }
            }
        }

        return ignored;
    }

    private bool MatchesPattern(string filePath, string[] pathParts, IgnorePattern pattern, bool isDirectory)
    {
        // Check if pattern applies to this file based on path prefix
        if (pattern.PathPrefix != ".")
        {
            var prefixParts = pattern.PathPrefix.Split('/');
            if (pathParts.Length < prefixParts.Length)
                return false;

            for (int i = 0; i < prefixParts.Length; i++)
            {
                if (prefixParts[i] != pathParts[i])
                    return false;
            }
        }

        // Get the relative path from the .gitignore location
        string relativePath;
        if (pattern.PathPrefix == ".")
        {
            relativePath = filePath;
        }
        else
        {
            relativePath = filePath.Substring(pattern.PathPrefix.Length + 1);
        }

        // If directory-only pattern, only match directories
        if (pattern.IsDirectoryOnly && !isDirectory)
        {
            return false;
        }

        return MatchPatternString(relativePath, pathParts, pattern.Pattern);
    }

    private bool MatchPatternString(string filePath, string[] pathParts, string pattern)
    {
        // Handle patterns with /
        if (pattern.Contains('/'))
        {
            // Pattern with / is anchored
            var regexPattern = pattern;
            if (pattern.StartsWith("/"))
            {
                regexPattern = pattern.Substring(1);
            }

            // Handle ** (matches zero or more directories)
            regexPattern = regexPattern.Replace("**", "<<<DOUBLESTAR>>>");

            // Handle * (matches anything except /)
            regexPattern = regexPattern.Replace("*", "[^/]*");

            // Handle ? (matches single character except /)
            regexPattern = regexPattern.Replace("?", "[^/]");

            // Restore ** as .*
            regexPattern = regexPattern.Replace("<<<DOUBLESTAR>>>", ".*");

            // Escape other regex special characters
            regexPattern = EscapeRegexExcept(regexPattern, "[^/].*$+()|");

            // Match pattern
            var regex = new System.Text.RegularExpressions.Regex($"^{regexPattern}(/.*)?$");
            return regex.IsMatch(filePath);
        }
        else
        {
            // Pattern without / matches at any depth
            var escapedPattern = EscapeRegex(pattern).Replace("\\*\\*", ".*");
            escapedPattern = escapedPattern.Replace("\\*", "[^/]*");
            escapedPattern = escapedPattern.Replace("\\?", ".");

            var regex = new System.Text.RegularExpressions.Regex($"(^|/){escapedPattern}$");
            return regex.IsMatch(filePath);
        }
    }

    private string EscapeRegex(string str)
    {
        var result = "";
        var specialChars = ".*+?^${}()|[]\\";
        foreach (var c in str)
        {
            if (specialChars.Contains(c))
            {
                result += "\\" + c;
            }
            else
            {
                result += c;
            }
        }
        return result;
    }

    private string EscapeRegexExcept(string str, string except)
    {
        var exceptSet = new HashSet<char>(except);
        var result = "";
        var specialChars = ".*+?^${}()|[]\\";
        foreach (var c in str)
        {
            if (specialChars.Contains(c) && !exceptSet.Contains(c))
            {
                result += "\\" + c;
            }
            else
            {
                result += c;
            }
        }
        return result;
    }

    private string? GetRelativePath(string basePath, string targetPath)
    {
        if (!targetPath.StartsWith(basePath))
        {
            return ".";
        }

        var baseUri = new Uri(basePath.EndsWith(Path.DirectorySeparatorChar.ToString()) ? basePath : basePath + Path.DirectorySeparatorChar);
        var targetUri = new Uri(targetPath.EndsWith(Path.DirectorySeparatorChar.ToString()) ? targetPath : targetPath + Path.DirectorySeparatorChar);
        var relativeUri = baseUri.MakeRelativeUri(targetUri);
        var relativePath = Uri.UnescapeDataString(relativeUri.ToString()).Replace('/', Path.DirectorySeparatorChar);

        return string.IsNullOrEmpty(relativePath) ? "." : relativePath.TrimEnd(Path.DirectorySeparatorChar);
    }

    // Test helper
    public void SetPatternsForTesting(Dictionary<string, List<IgnorePattern>> testPatterns)
    {
        _patterns = testPatterns;
    }
}
